README.md
[![Build status — Travis-CI][travis-icon]][travis]
[![cookie-notice on Npmjs][npm-icon]][npm]
[![License][license-icon]][mit]
[![Total downloads ~ Npmjs][downl-icon]][npm]
[![Size of Javascript][size-icon]][build]

# CookieNotice

**CookieNoticeJS** is a very simple and small *(→ 2 kB gzip)* vanilla JS script with multi language support for [GDPR][]/[DSGVO][]‎ transparency and
notification purposes that provides an easy way to show a cookie notice on your website.

<img src="https://i.imgur.com/koDf1h0.png" alt="cookie notice working example" align="center"/>

**Available via [npm][]**

```shell
npm install cookie-notice
npm test
```


**To use in your project There are plenty ways for integration:**

**When installed via npm, include in any project by using path below:**
```html
<script src="node-modules/cookie-notice/dist/cookie.notice.min.js"></script>
```

**For usage with Angular2+ add line below in "scripts" section in angular-cli.json:**
```json
{
    "scripts": [
        "../node_modules/cookie-notice/dist/cookie.notice.js",
        "../optional/path/to/custom/cookie-notice-config.js"
    ]
}
```

**When cloned directly from gitHub use path below:**
```html
<script src="cookie-notice/dist/cookie.notice.min.js"></script>
```

## Behavior
You will get a dismissable banner on the bottom of your pages showing a default cookie audit like the following:

> We use cookies to make sure you can have the best experience on our website. If you continue to use this site we assume that you will be happy with it.

Check my website for a [DEMO](http://codeb.it/). Depending on the visitor browser language one of the preloaded translations will be shown. At the moment **CookieNoticeJS** supports *EN, IT, DE and FR*. If you want to contribute with an extra language do not hesitate to open an issue or a PR.

**CookieNoticeJS** has been successfully tested on IE9+, Chrome, Firefox and Safari.

## Customize CookieNoticeJS

For the most of you including the script should be enough but **CookieNoticeJS** comes with many customization options. Let's see an example:

```html
<script src="js/cookie.notice.min.js"></script>
<script>
    new cookieNoticeJS({

       // Localizations of the notice message
       'messageLocales': {
         'it': 'Custom localized message'
       },

       // Localizations of the dismiss button text
       'buttonLocales': {
         'it': 'Chiudi'
       },

       // Position for the cookie-notifier (default=bottom)
       'cookieNoticePosition': 'top',

       // Shows the "learn more button (default=false)
       'learnMoreLinkEnabled': false,

       // The href of the learn more link must be applied if (learnMoreLinkEnabled=true)
       'learnMoreLinkHref': '/learn/more/index.html',

       // Text for optional learn more button
       'learnMoreLinkText':{
           'en':'learn more'
       },

       // The message will be shown again in X days
       'expiresIn': 30,

      // Specify a custom font family and size in pixels
      'fontFamily': 'inherit',
      'fontSize': 12,

       // Dismiss button background color
       'buttonBgColor': '#d35400',  

       // Dismiss button text color
       'buttonTextColor': '#fff',

       // Notice background color
       'noticeBgColor': '#000',

       // Notice text color
       'noticeTextColor': '#fff',

       // the learnMoreLink color (default='#009fdd')
       'linkColor': '#f00',

       // The target of the learn more link (default='', or '_blank')
       'linkTarget': '',

       // Print debug output to the console (default=false)
       'debug': false
    });
</script>
```

## Configuration via `data-` attribute

Configuration options can be put in a `data-cookie-notice` HTML attribute in JSON format.
Note, you can include the Javascript from the [unpkg][] CDN ([browse][]).
For example:

```html
<script
  data-cookie-notice='{ "learnMoreLinkEnabled": true, "learnMoreLinkHref": "/privacy.html" }'
  src="https://unpkg.com/cookie-notice@^1/dist/cookie.notice.min.js"
></script>
```

#### Author
Alessandro Benoit

#### Contributors
[Bernhard Behrendt](mailto:bernhard.behrendt@aoe.com) [@AOEpeople](https://github.com/AOEpeople),
[Nick Freear](https://github.com/nfreear) [IET at the OU](https://github.com/IET-OU)

### License

License: [MIT][]


[MIT]: https://github.com/micc83/cookie-notice-js/blob/master/LICENSE
    "License: MIT | Copyright © 2018 Alessandro Benoit (micc83)."
[MIT-0]: https://mit-license.org/#2018
[travis-icon]: https://travis-ci.org/AOEpeople/cookie-notice.svg?branch=master
[travis]: https://travis-ci.org/AOEpeople/cookie-notice "Build status – Travis-CI"
[npm]: https://npmjs.com/package/cookie-notice "CookieNotice – on NPM"
[npm-icon]: https://badge.fury.io/js/cookie-notice.svg
[npm-i0]: https://img.shields.io/npm/v/cookie-notice.svg "(Timeout errors)"
[license-icon]: https://img.shields.io/npm/l/cookie-notice.svg
[downl-icon]: https://img.shields.io/npm/dt/cookie-notice.svg "Count of total downloads – NPM"
[build]: https://github.com/AOEpeople/cookie-notice/tree/master/dist
[size-icon]: https://img.shields.io/github/size/AOEpeople/cookie-notice/dist/cookie.notice.min.js.svg
    "Size of built Javascript, kilo-bytes (kB) – on GitHub"
[unpkg]: https://unpkg.com/ "unpkg is a fast, global content delivery network (CDN) for everything on npm."
[browse]: https://unpkg.com/cookie-notice@^1/ "Browse cookie-notice on unpkg"

[DSGVO]: https://de.wikipedia.org/wiki/Datenschutz-Grundverordnung "Datenschutz-Grundverordnung (DSGVO)"
[GDPR]: https://en.wikipedia.org/wiki/General_Data_Protection_Regulation "General Data Protection Regulation (GDPR)"

[End]: //.


CHANGELOG.md

# CookieNotice changelog

← [README][]

Version History

## Version 1.1.19
 * _Date: 09 July 2020_
 * Added fontSize customization, with fallback to 12px if unset

## Version 1.1.18
 * _Date:  01 January 2019_;
 * Step back to px based font sizes due to negative side effects

 ## Version 1.1.17
 * _Date:  15 December 2018_;
 * Accessibility (a11y) improvements - fixed all WCAG2AA errors and warnings

## Version 1.1.16
 * _Date:  12 December 2018_;
 * Initial font size too small

 ## Version 1.1.14/15
 * _Date:  6 December 2018_;
 * Typo in German translation -> contrib. by @chroedermatthias (Matthias Schröder)
 * Dependency upgrade
 * Fixed unit tests

## Version 1.1.12

 * _Date:  approx. 4-6 June 2018_;
 * Add accessibility testing via [grunt-rsids-pa11y][] and [pa11y][], WCAG 2.0 AA, pull #6;
 * Accessibility fixes — colour contrast, button role, link target, link underline;
 * Accessibility fixes — express CSS font-size in `em` or `rem`, not `px`;
 * Add configuration option — `linkTarget` (default: '');
 * Edit the language used in the English translation;
 * Fix `browser` and `repository` fields in `package.json`;
 * Add `files` and `bugs` fields to `package.json`;
 * Add `engines` and `contributors` fields to `package.json`;
 * Add `live-server` to `peerDependencies` in `package.json`;
 * Remove Nick's email address from `README` & `bower.json` (!)

## Version 1.1.11

 * _Date:  30 May 2018_;
 * Add support for `data-cookie-notice` attribute on `<script>`, pull #5, issue [#4][];
 * Document `data-` attribute in README;
 * Document _unpkg_ CDN in README;

## Version 1.1.10

 * _Date:  29 May 2018_;
 * Move jQuery to `devDependencies` for clarity, pull #2;
 * Point `tests/index.html` to local copies of jQuery and QUnit (was CDN);

## Version 1.1.9

 * _Date:  25 May 2018_;
 * Add Travis-CI testing to the project, pull #1;
 * Update test dependencies —
   jQuery from 2.x → 3.x; QUnit from 1.x → 2.x;
 * Add SVG badges for Travis-CI, NPMJS.com, etc. to `README.md`;

## Version 1.1.8

 * _Date:  25 May 2018_;

## Version 1.1.7

 * _Date:  25 April 2018_;

...

## Version 1.1.0

 * _Date:  24 April 2018_;
 * First release on [npm][];

## Version 1.0.0

 * _Date:  20 May 2015_;
 * Release by:  [@micc83][];
 * (Not on [npm][])

---

← [README][]

[readme]: https://github.com/AOEpeople/cookie-notice#readme
[npm]: https://www.npmjs.com/package/cookie-notice
[@micc83]: https://github.com/micc83/cookie-notice-js "Alessandro Benoit"
[#4]: https://github.com/AOEpeople/cookie-notice/issues/4
[a11y-fail]: https://travis-ci.org/nfreear/cookie-notice/jobs/387344672#L1217-L1235
[grunt-rsids-pa11y]: https://www.npmjs.com/package/grunt-rsids-pa11y
[pa11y]: http://pa11y.org/


src/cookie.notice.js
/**
 * Cookie Notice JS
 * @author Alessandro Benoit
 */
(function () {

    "use strict";

    /**
     * Store current instance
     */
    var instance,
        originPaddingTop;

    /**
     * Defaults values
     * @type object
     */
    var defaults = {
        messageLocales: {
            it: 'Utilizziamo i cookie per essere sicuri che tu possa avere la migliore esperienza sul nostro sito. Se continui ad utilizzare questo sito assumiamo che tu ne sia felice.',
            en: 'We use cookies to ensure that you have the best experience on our website. If you continue to use this site we assume that you accept this.',
            de: 'Wir verwenden Cookies um sicherzustellen, dass Sie das beste Erlebnis auf unserer Website haben.',
            fr: 'Nous utilisons des cookies afin d\'être sûr que vous pouvez avoir la meilleure expérience sur notre site. Si vous continuez à utiliser ce site, nous supposons que vous acceptez.',
            pt: 'Utilizamos cookies para garantir que você tenha a melhor experiência em nosso site. Se você continuar a usar este site, assumimos que você aceita isso.'
        },

        cookieNoticePosition: 'bottom',

        learnMoreLinkEnabled: false,

        learnMoreLinkHref: '/cookie-banner-information.html',

        learnMoreLinkText: {
            it: 'Saperne di più',
            en: 'Learn more',
            de: 'Mehr erfahren',
            fr: 'En savoir plus',
            pt: 'Saber mais'
        },

        buttonLocales: {
            en: 'OK'
        },

        expiresIn: 30,

        fontFamily: 'inherit',
        fontSize: 12,

        buttonBgColor: '#ca5000',
        buttonTextColor: '#fff',
        noticeBgColor: '#000',
        noticeTextColor: '#fff',
        linkColor: '#009fdd',
        linkBgColor: '#000',
        linkTarget: '_blank',
        debug: false
    };

    /**
     * Initialize cookie notice on DOMContentLoaded
     * if not already initialized with alt params
     */
    document.addEventListener('DOMContentLoaded', function () {
        if (!instance) {
            new cookieNoticeJS();
        }
    });

    /**
     * Constructor
     * @constructor
     */
    window.cookieNoticeJS = function () {

        // If an instance is already set stop here
        if (instance !== undefined) {
            return;
        }

        // Set current instance
        instance = this;

        // If cookies are not supported or notice cookie is already set
        if (getNoticeCookie()) {
            return;
        }

        // 'data-' attribute - data-cookie-notice='{ "key": "value", ... }'
        var elemCfg = document.querySelector('script[ data-cookie-notice ]');
        var config;

        try {
            config = elemCfg ? JSON.parse(elemCfg.getAttribute('data-cookie-notice')) : {};
            // TODO apply settings coming from data attribute and keep defaults if not overwritten -> 1.2.x
        } catch (ex) {
            console.error('data-cookie-notice JSON error:', elemCfg, ex);
            config = {};
        }

        // Extend default params
        var params = extendDefaults(defaults, arguments[0] || config || {});

        if (params.debug) {
            console.warn('cookie-notice:', params);
        }

        // Get current locale for notice text
        var noticeText = getStringForCurrentLocale(params.messageLocales);

        // Create notice
        var notice = createNotice(noticeText, params.noticeBgColor, params.noticeTextColor, params.fontFamily, params.fontSize, params.cookieNoticePosition);

        var learnMoreLink;

        if (params.learnMoreLinkEnabled) {
            var learnMoreLinkText = getStringForCurrentLocale(params.learnMoreLinkText);

            learnMoreLink = createLearnMoreLink(learnMoreLinkText, params.learnMoreLinkHref, params.linkTarget, params.linkColor, params.linkBgColor);
        }

        // Get current locale for button text
        var buttonText = getStringForCurrentLocale(params.buttonLocales);

        // Create dismiss button
        var dismissButton = createDismissButton(buttonText, params.buttonBgColor, params.buttonTextColor, params.fontFamily);

        // Dismiss button click event
        dismissButton.addEventListener('click', function (e) {
            e.preventDefault();
            setDismissNoticeCookie(parseInt(params.expiresIn + "", 10) * 60 * 1000 * 60 * 24);
            fadeElementOut(notice);
        });

        // Append notice to the DOM
        var noticeDomElement = document.body.appendChild(notice);

        if (!!learnMoreLink) {
            noticeDomElement.appendChild(learnMoreLink);
        }

        noticeDomElement.appendChild(dismissButton);

    };

    /**
     * Get the string for the current locale
     * and fallback to "en" if none provided
     * @param locales
     * @returns {*}
     */
    function getStringForCurrentLocale(locales) {
        var locale = (navigator.userLanguage || navigator.language).substr(0, 2);
        return (locales[locale]) ? locales[locale] : locales['en'];
    }

    /**
     * Test if notice cookie is there
     * @returns {boolean}
     */
    function getNoticeCookie() {
        return document.cookie.indexOf('cookie_notice') != -1;
    }

    /**
     * Create notice
     * @param message
     * @param bgColor
     * @param textColor
     * @param position
     * @param fontFamily
     * @returns {HTMLElement}
     */
    function createNotice(message, bgColor, textColor, fontFamily, fontSize, position) {
        var notice = document.createElement('div'),
            noticeStyle = notice.style,
            lineHeight = 28,
            paddingBottomTop = 10,
            noticeHeight = lineHeight + paddingBottomTop * 2;
        fontSize = typeof fontSize !== 'undefined' ? fontSize : 12;

        notice.innerHTML = message + '&nbsp;';
        notice.setAttribute('id', 'cookieNotice');
        notice.setAttribute('data-test-section', 'cookie-notice');
        notice.setAttribute('data-test-transitioning', 'false');


        noticeStyle.position = 'fixed';

        if (position === 'top') {
            var bodyDOMElement = document.querySelector('body');

            originPaddingTop = bodyDOMElement.style.paddingTop;

            noticeStyle.top = '0';
            bodyDOMElement.style.paddingTop = noticeHeight + 'px';
        } else {
            noticeStyle.bottom = '0';
        }


        noticeStyle.left = '0';
        noticeStyle.right = '0';
        noticeStyle.background = bgColor;
        noticeStyle.color = textColor;
        noticeStyle["z-index"] = '999';
        noticeStyle.padding = paddingBottomTop + 'px 5px';
        noticeStyle["text-align"] = 'center';
        noticeStyle["font-size"] = fontSize + 'px';
        noticeStyle["line-height"] = lineHeight + 'px';

        if (!!fontFamily) {
            noticeStyle['fontFamily'] = fontFamily;
        }

        return notice;
    }

    /**
     * Create dismiss button
     * @param message
     * @param buttonColor
     * @param buttonTextColor
     * @returns {HTMLElement}
     */
    function createDismissButton(message, buttonColor, buttonTextColor, buttonTextFontFamily) {

        var dismissButton = document.createElement('span'),
            dismissButtonStyle = dismissButton.style;

        // Dismiss button
        dismissButton.href = '#';
        dismissButton.innerHTML = message;

        dismissButton.setAttribute('role', 'button');
        dismissButton.className = 'confirm';

        dismissButton.setAttribute('data-test-action', 'dismiss-cookie-notice');


        // Dismiss button style
        dismissButtonStyle.background = buttonColor;
        dismissButtonStyle.color = buttonTextColor;
        dismissButtonStyle['text-decoration'] = 'none';
        dismissButtonStyle['cursor'] = 'pointer';
        dismissButtonStyle.display = 'inline-block';
        dismissButtonStyle.padding = '0 15px';
        dismissButtonStyle.margin = '0 0 0 10px';

        if (!!buttonTextFontFamily) {
            dismissButtonStyle.fontFamily = buttonTextFontFamily;

        }

        return dismissButton;

    }

    /**
     * Create the learn more link
     *
     * @param learnMoreLinkText
     * @param learnMoreLinkHref
     * @param linkColor
     * @returns {HTMLElement}
     */
    function createLearnMoreLink(learnMoreLinkText, learnMoreLinkHref, linkTarget, linkColor, linkBgColor) {

        var learnMoreLink = document.createElement('a'),
            learnMoreLinkStyle = learnMoreLink.style;

        learnMoreLink.href = learnMoreLinkHref;
        learnMoreLink.textContent = learnMoreLinkText;
        learnMoreLink.title = learnMoreLinkText;
        learnMoreLink.target = linkTarget;
        learnMoreLink.className = 'learn-more';
        learnMoreLink.setAttribute('data-test-action', 'learn-more-link');

        learnMoreLinkStyle.color = linkColor;
        learnMoreLinkStyle.backgroundColor = 'transparent';
        learnMoreLinkStyle['text-decoration'] = 'underline';
        learnMoreLinkStyle.display = 'inline';

        return learnMoreLink;

    }

    /**
     * Set sismiss notice cookie
     * @param expireIn
     */
    function setDismissNoticeCookie(expireIn) {
        var now = new Date(),
            cookieExpire = new Date();

        cookieExpire.setTime(now.getTime() + expireIn);
        document.cookie = "cookie_notice=1; expires=" + cookieExpire.toUTCString() + "; path=/;";
    }

    /**
     * Fade a given element out
     * @param element
     */
    function fadeElementOut(element) {
        element.style.opacity = 1;

        element.setAttribute('data-test-transitioning', 'true');

        (function fade() {
            if ((element.style.opacity -= .1) < 0.01) {

                if (originPaddingTop !== undefined) {
                    var bodyDOMElement = document.querySelector('body');
                    bodyDOMElement.style.paddingTop = originPaddingTop;
                }

                document.body.removeChild(element);
            } else {
                setTimeout(fade, 40);
            }
        })();
    }

    /**
     * Utility method to extend defaults with user options
     * @param source
     * @param properties
     * @returns {*}
     */
    function extendDefaults(source, properties) {
        var property;
        for (property in properties) {
            if (properties.hasOwnProperty(property)) {
                if (typeof source[property] === 'object') {
                    source[property] = extendDefaults(source[property], properties[property]);
                } else {
                    source[property] = properties[property];
                }
            }
        }
        return source;
    }

    /* test-code */
    cookieNoticeJS.extendDefaults = extendDefaults;
    cookieNoticeJS.clearInstance = function () {
        instance = undefined;
    };
    /* end-test-code */

}());
